module/zfs/zio.c

```diff
@@ -3245,16 +3245,36 @@ zio_ddt_write(zio_t *zio)
 	ddt_t *ddt = ddt_select(spa, bp);
 	ddt_entry_t *dde;
 	ddt_phys_t *ddp;
+	boolean_t added = B_FALSE;
 
 	ASSERT(BP_GET_DEDUP(bp));
 	ASSERT(BP_GET_CHECKSUM(bp) == zp->zp_checksum);
 	ASSERT(BP_IS_HOLE(bp) || zio->io_bp_override);
 	ASSERT(!(zio->io_bp_override && (zio->io_flags & ZIO_FLAG_RAW)));
 
 	ddt_enter(ddt);
-	dde = ddt_lookup(ddt, bp, B_TRUE);
+	if (spa->spa_dedup_max_entries &&
+	    spa->spa_dedup_entries >= spa->spa_dedup_max_entries) {
+		/* Over the desired DDT size, set 'nogrow' flag */
+		dde = ddt_lookup(ddt, bp, B_TRUE, NULL);
+	} else {
+		dde = ddt_lookup(ddt, bp, B_FALSE, &added);
+	}
+	if (dde == NULL) {
+		zp->zp_dedup = B_FALSE;
+		BP_SET_DEDUP(bp, B_FALSE);
+		ASSERT(!BP_GET_DEDUP(bp));
+		zio->io_pipeline = ZIO_WRITE_PIPELINE;
+		ddt_exit(ddt);
+		return (zio);
+	}
+
 	ddp = &dde->dde_phys[p];
 
+	if (added) {
+		spa->spa_dedup_entries++;
```

Comment:
So this is the actual number of entries (blocks) in the DDT. We increment it here to avoid going over the limit until we update with the accurate count during ddt_sync()

I suppose we could decrement it in zio_ddt_free(), but I am not sure it is worth it

Reply:
When a dedup'd dataset gets destroyed, all the refcounts go down in the sync, and then as part of the ddt_sync(), the counts get recomputed.